pkg/query-service/app/preferences/model.go

```diff
@@ -0,0 +1,598 @@
+	preferenceValue := PreferenceKV{PreferenceId: preferenceId, PreferenceValue: preference.DefaultValue}
+
+	if preference.OrgScope == 1 && orgPreference.PreferenceId != "" {
+		preferenceValue.PreferenceValue = orgPreference.PreferenceValue
+	}
+
+	if userPreference.PreferenceId != "" {
+		preferenceValue.PreferenceValue = userPreference.PreferenceValue
+	}
+
+	return &preferenceValue, nil
+}
+
+func UpdateUserPreference(ctx context.Context, req *PreferenceKV) (*PreferenceKV, *model.ApiError) {
+	preferenceId := req.PreferenceId
+	preferenceValue := req.PreferenceValue
+	user := common.GetUserFromContext(ctx)
+
+	// return error if there is no preference id in the request
+	if preferenceId == "" {
+		return nil, &model.ApiError{Typ: model.ErrorNotFound, Err: fmt.Errorf("no preference id found in the request")}
+	}
+
+	preference := Preference{}
+	query := `SELECT user FROM preference WHERE id=$1;`
+	err := db.Get(&preference, query, preferenceId)
+
+	if err != nil {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("no such preference exists: %s", err)}
+	}
+
+	if preference.UserScope != 1 {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("this preference is not enabled at user scope: %s", preferenceId)}
+	}
+
+	query = `INSERT INTO user_preference(preference_id,preference_value,user_id) VALUES($1,$2,$3)
+	ON CONFLICT(preference_id,user_id) DO
+	UPDATE SET preference_value= $2 WHERE preference_id=$1 AND user_id=$3;
+	`
+
+	_, err = db.Exec(query, preferenceId, preferenceValue, user.Id)
+
+	if err != nil {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in setting the preference value: %s", err)}
+	}
+
+	return &PreferenceKV{
+		PreferenceId:    preferenceId,
+		PreferenceValue: preferenceValue,
+	}, nil
+}
+
+func GetOrgPreference(ctx context.Context, preferenceId string, orgId string) (*PreferenceKV, *model.ApiError) {
+	orgPreference := PreferenceKV{}
+
+	// get the preference id and value from the org preference table
+	query := `SELECT preference_id, preference_value FROM org_preference WHERE preference_id = $1 AND org_id = $2;`
+	err := db.Get(&orgPreference, query, preferenceId, orgId)
+
+	if err != nil && err != sql.ErrNoRows {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: err}
+	}
+
+	// get the details for the preference entity
+	preference := Preference{}
+	query = `SELECT org, default_value FROM preference WHERE id = $1;`
+	err = db.Get(&preference, query, preferenceId)
+
+	// return if unable to fetch the preference entity as we won't be sure about preference being enabled or not
+	if err != nil {
+		return nil, &model.ApiError{Typ: model.ErrorNotFound, Err: fmt.Errorf("error while fetching the preference entity: %s", preferenceId)}
+	}
+
+	// return err if the preference is not enabled for org scope
+	if preference.OrgScope != 1 {
+		return nil, &model.ApiError{Typ: model.ErrorNotFound, Err: fmt.Errorf("preference not enabled with key: %s", preferenceId)}
```

Comment:
This should be bad request rather than not found?

Reply:
done